### ISSUE TITLE
Refactor app.js

### ISSUE THREAD

**Issue by yann300** [ethereum/remix-ide#1261]
Locally trigger event and globally register

```
compiler.event.trigger('compilationFinished',` [])
compiler.id = 'compilerID'

globalEvent.registerEvents(compiler.event)

event = new EventManager()
registerEvents = (event) => {
    forEach(event) {
     registerTo(event)
    }
}

// globally register
globalEventRegister.register('compilerID.compilationFinished', (args) => {})
```

**Comment by serapath** [comment]
maybe this can be joined into one thing?
```js
// MODULE 1
var registry = require('global-inter-communication-registry')

module.exports = class Compiler {
  constructor (opts) {
    const self = this


    // 1. compiler publishes api and events
    var { api, event } = registry.publish('compilerID')
    // 1.1. trigger events
    event.trigger('hello', ['world'])
    // 1.3. publish api
    api.publish('methodName', (...args) => self.methodname.apply(self, args))

    // 2. compiler registers to other events and uses other api
    var { api: terminalAPI, event: terminalEVENTS } = registry.install('terminalID')
    // 2.1 register to events
    terminalEVENTS.register('foo', event => console.log(event))
    // 2.2. call api - sync method
    var result = terminalAPI.doSyncStuff(1, 2, 3)
    console.log('sync', result)
    // 2.3. call api - async method
    terminalAPI.doAsyncStuff(1, 2, 3, result => console.log('async', result))


  }
  compile (source) { /* ... */ }
  switchVersion (version, callback) { /* ... */ }
}
```

```js
// MODULE 2
var registry = require('global-inter-communication-registry')

module.exports = class Terminal {
  constructor (opts) {
    const self = this


    // 1. terminal can publishe it's own api and events
    var { api, event } = registry.publish('terminalID')
    // 1.1. announce events
    var triggerFoo = event('foo')
    // 1.2. emit events
    triggerfoo('bar') // THINK: self.event.trigger('foo', ['bar'])
    // 1.3. publish api
    api('doSyncStuff', (a, b, c) => self.doSyncStuff(a, b, c))
    api('doAsyncStuff', (a, b, c, done) => self.doAsyncStuff(a, b, c, done))

    // 2. terminal registers to other events and uses other apis
    var { api: compilerAPI, event: compilerEVENTS } = registry.install('compilerID')
    // 2.1 register to events
    compilerEVENTS('hello', event => console.log(event))
    // 2.2. call api - sync method
    var result = compilerAPI.compile(`pragma solidity ^0.4.0; contract Ballot { }`)
    console.log('sync', result)
    // 2.3. call api - async method
    compilerAPI.switchVersion('v1.2.3', () => console.log('success'))


  }
  doSyncStuff (a, b, c) { /* ... */ }
  doAsyncStuff (a, b, c, done) { /* ... */ }
}
```

**Comment by yann300** [comment]
Some comments:
 - Finally I am not sure if publishing/installing using `<some id>` is good, because we will then need to ensure ids are unique and that could be source of bug.
 - Returning both event and API from `registry.install('module1 ID')` is nice and convenient
 - Defining the API like `api('doSyncStuff', (a, b, c) => self.doSyncStuff(a, b, c))` would be quite verbose and not very efficient.  I would perhaps first just make the public function available through the API without having a special obj for that.
 - the following is related to the previous comment ^ : I think it makes sense at this stage to think about how to restrict APIs rather than what we want to make available. I mean that:
       - all the instances have public methods (API available through the entire app)
       - if a plugin want to use a function of a specific instance, he would have to declare it in a permission manifest.json like: `{ 'compiler' : [ 'getCompilationData' , 'compile' ] }`

**Comment by serapath** [comment]
1. unique id's for publish/install
    1. the "registry" could throw if an ID is used more than once, so you would immediately know and fix it
    2. we can use the little stacktrace trick to auto-generate the IDs to solve the problem
2. great :-)
3. not sure what you mean with "verbose" regarding `api('doSyncStuff', ...)`
    1. we could do `api(self, 'doSyncStuff')` so `api()` knows the method name to call on self maybe?
    2. or even just `api(selfProxy)` which has only important messages. is this how you imagine?
4. i agree. if you think about (3.1.) ...this has already "permissions" in mind, but this is a very good point

I will think about it and post again bit later

**Comment by yann300** [comment]
The stack trace util could be useful at some point yes. I agree with yout, then let's put the permission layer after ... but soon :)

I think there should be two kind of registry, one for our internal stuff (where ids would be like 'compiler' or 'udapp').
internal module would be published that way: 
`{ event } = registry.publish(moduleId, this)` and later replace by 
`{ event } = registry.publish(moduleId, selfProxy)`
first param being taken from the module constructor `constructor (moduleId, opts) {`
second param being the module itself.

I intentionally make it return `{ event }`, so we can be flexible.
So basically all the functions associated with the module would be part of the API.
We'll add more complexity later ( like permissionning layer and API availability ).

For plugin registries that look the same except that `moduleId` would be something like 'name/45ad` <commonname/hex id>
We will need to specify how to plug this with `postMessage` but that'll be another topic...

**Comment by yann300** [comment]
@serapath 
to resume, we would have: 
`{ event } = registry.publish(moduleId, this)` to register a new module
and
`{ api, event } = registry.get('moduleId')` to use an existing module

**Comment by serapath** [comment]
I will try to put all of this and comments in another code example like the one above, where this time, i would have many modules using our module, that i will just call `appcomm` here.

There would be modules that communicate with each other in hopefully all possible ways, maybe including an iframe and one module (maybe `app.js`) that will log all communication that is going on, which is exactly the logic we would use to debug.

I feel the ideas we had in the plugin API interface should be re-used in parts  to make it more easily compatible.
Of course - like the iframe api should or could have a helper module to make it more convenient to use, the same way the `appcomm` module could have wrapper to make it convenient to use.

This sounds abstract, so in the next post i will try to post the example code, but i need some time to prepare it

**Comment by serapath** [comment]
```js
var registry = require('registry')

var myEVENTS = [ 'a', 'b', 'c']
var myAPI = {
  foo () { console.log("somebody called my method foo")
  bar () { console.log("somebody called my method bar")
}

var { uid, api, events: { a, b, c } = registry.publish({ api: myAPI, events: myEVENTS })
console.log(uid) // ... e.g. "app/righthandpanel/settings/123" or something that is a unique identifier

var data = "foobar"
// trigger events
a(data)
b(data)
c(data)

// publish additional event
var d = events("d")

// trigger new event
d(data)

// change published api
var myNewAPI { baz () { console.log("somebody called my baz method") } }
api(myNewAPI)
```

**Comment by serapath** [comment]
# consumer
```js
var registry = require('registry')

setupB(registriy.install("app/righthandpanel/settingstab/123"))
setupC(registry.install("app/righthandpanel/compiltab/234"))

async function setupB ({ uid, api, events }) {
  events('a', event => console.log(event)) // => "foobar"
  var result = await api.foo(123)
  console.log('sync', result) // => true
  api.bar(123, result => console.log('async', result)) // => "barbaz"
}

async function setupC ({ uid, api, events }) { /* ... */ }

```

**Comment by serapath** [comment]
# devtools
i'm not yet sure how the devtools should look like. my current take on it is something like:

```js
var registry = require('registry')

registry.devtools.onregister(({ uid, api, events }, done) => {
  console.log('register', uid)
  done()
})
registry.devtools.oninstall(({ puid, suid }, done) => {
  console.log('publisher', puid, 'subscriber', suid)
  done()
})
registry.devtools.onpublish(({ uid, type, data }, done) => {
  console.log(uid, type === 'api' ? 'api' : 'events', data)
  done()
})
register.devtools.onsubscribe(({ suid, puid, name }, done) => {
  console.log('subscriber', suid, 'eventname', name, 'emitter', puid)
  done()
})
register.devtools.onmessage({ uid1, uid2, name, data }, done) => {
  console.log('sender, suid, 'receiver', puid, 'method/event', name, 'args/event', data)
  done()
})
```

**Comment by yann300** [comment]
```
var { uid, api: publishAPI, events: publishEVENTS } = registry.put({ api, events, name })
var { uid, api: useAPI, events: subscribeEVENTS } = registriy.get(uid) // uid = "ip" or "domain"
```
